lib/guess_html_encoding.rb

```diff
@@ -19,11 +19,20 @@ def self.guess(html, headers = nil)
     end
 
     if out.nil? || out.empty? || !encoding_loaded?(out)
-      if html =~ /<meta[^>]*HTTP-EQUIV=["']?Content-Type["']?[^>]*content=["']([^'"]*)["']/i && $1 =~ /charset=([\w\d-]+);?/i
-        out = $1
-      elsif html =~ /<meta\s+charset=["']([\w\d-]+)?/i
-        out = $1
+
+      meta_tags_regex = /
+      (?:
+        <meta[^>]*HTTP-EQUIV=["']?Content-Type["']?[^>]*content=["'][^'"]*charset\s*=\s*([^"'\;\s]+)  # http-equiv
+        |                                                                                              # or
```

Comment:
Agreed, this could definitely be more robust.  I'm happy to take additional pull requests, and / or to slowly improvement it over time myself.